ddl/db_test.go

```diff
@@ -1470,3 +1470,38 @@ func TestReportingMinStartTimestamp(t *testing.T) {
 	infoSyncer.ReportMinStartTS(dom.Store())
 	require.Equal(t, validTS, infoSyncer.GetMinStartTS())
 }
+
+// for issue #34931
+func TestBuildMaxLengthIndexWithNonRestrictedSqlMode(t *testing.T) {
+	store, clean := testkit.CreateMockStore(t)
+	defer clean()
+
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+
+	tk.MustExec("drop table if exists t1")
+	tk.MustExec("drop table if exists t2")
+	tk.MustExec("drop table if exists t3")
+
+	// test in strict sql mode
+	sql := "create table t1 (id int, name varchar(2048), index(name)) charset=utf8;"
+	tk.MustGetErrCode(sql, errno.ErrTooLongKey)
+
+	tk.MustExec("set @@sql_mode=''")
+
+	err := tk.ExecToErr("create table t2 (id int, name varchar(2048), index(name)) charset=utf8;")
+	require.NoError(t, err)
+	require.Equal(t, uint16(1), tk.Session().GetSessionVars().StmtCtx.WarningCount())
+	tk.MustQuery("show create table t2").Check(testkit.Rows("t2 CREATE TABLE `t2` (\n  `id` int(11) DEFAULT NULL,\n  `name` varchar(2048) DEFAULT NULL,\n  KEY `name` (`name`(1024))\n) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_bin"))
+
+	// For a unique index, an error occurs regardless of SQL mode because reducing
+	//the index length might enable insertion of nonunique entries that do not meet
+	//the specified uniqueness requirement.
+	sql = "create table t1 (id int, name varchar(2048), unique index(name)) charset=utf8;"
+	tk.MustGetErrCode(sql, errno.ErrTooLongKey)
+
+	// The multiple column index in which the length sum exceeds the maximum size
+	// will return an error instead produce a warning.
+	sql = "create table t3 (id int, name varchar(512), alias varchar(1024), index(name,alias)) charset=utf8;"
+	tk.MustGetErrCode(sql, errno.ErrTooLongKey)
```

Comment:
Please add the test case for more charsets:

* `ascii`, for which `colLenPerUint` is 1
* `utf8mb4`, for which `colLenPerUint` is 4

Also, please verify the result for `char`, which might(I'm not sure) be different with `varchar`.

Thank you!